In [1]:
%pip install numpy scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [2]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.naive_bayes import MultinomialNB, ComplementNB
from sklearn.metrics import f1_score

# 20newsgroups por ser un dataset clásico de NLP ya viene incluido y formateado
# en sklearn
from sklearn.datasets import fetch_20newsgroups
import numpy as np
import random as rnd
import pandas as pd
from IPython.display import display, HTML


### Consigna del desafío 1

**1**. Vectorizar documentos. Tomar 5 documentos al azar y medir similaridad con el resto de los documentos.
Estudiar los 5 documentos más similares de cada uno analizar si tiene sentido
la similaridad según el contenido del texto y la etiqueta de clasificación.

**2**. Construir un modelo de clasificación por prototipos (tipo zero-shot). Clasificar los documentos de un conjunto de test comparando cada uno con todos los de entrenamiento y asignar la clase al label del documento del conjunto de entrenamiento con mayor similaridad.

**3**. Entrenar modelos de clasificación Naïve Bayes para maximizar el desempeño de clasificación
(f1-score macro) en el conjunto de datos de test. Considerar cambiar parámteros
de instanciación del vectorizador y los modelos y probar modelos de Naïve Bayes Multinomial
y ComplementNB.

**4**. Transponer la matriz documento-término. De esa manera se obtiene una matriz
término-documento que puede ser interpretada como una colección de vectorización de palabras.
Estudiar ahora similaridad entre palabras tomando 5 palabras y estudiando sus 5 más similares. **La elección de palabras no debe ser al azar para evitar la aparición de términos poco interpretables, elegirlas "manualmente"**.


### Punto 1

#### Dataset 20 newsgroups, carga de datos

In [3]:
# cargamos los datos (ya separados de forma predeterminada en train y test)
newsgroups_train = fetch_20newsgroups(subset='train', remove=('headers', 'footers', 'quotes'))
newsgroups_test = fetch_20newsgroups(subset='test', remove=('headers', 'footers', 'quotes'))

#### Vectorización

In [4]:
# Vectorizo el dataset de train
tfidfvect = TfidfVectorizer()
X_train = tfidfvect.fit_transform(newsgroups_train.data)
y_train = newsgroups_train.target
# Vectorizo el dataset de test 
X_test = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

#### Similitud

In [5]:
rnd.seed(92)
# Tomo 5 muestras al azar, sobre los documentos vectorizados para el set de test.
doc_idxs = rnd.sample(range(0, X_test.shape[0]), 5)

for idx in doc_idxs:
    # Por cada muestra, busco los documentos similares
    cossim = cosine_similarity(X_test[idx], X_test)[0]
    # Ordeno los resultados de mayor a menor y me quedo con los 5 indices mas similares, ingorando el primero, que es el mismo documento
    mostsim = np.argsort(cossim)[::-1][1:6]
    list_sims = []
    for sim in mostsim:
        list_sims.append({"index": sim, "similitud": cossim[sim], "target": newsgroups_test.target_names[y_test[sim]], "texto": newsgroups_test.data[sim]})
    df_sims = pd.DataFrame(list_sims)
    print(f"Similitudes para documento: {idx}\nlabel: {newsgroups_test.target_names[y_test[idx]]}\ntexto:\n{newsgroups_test.data[idx]}")
    html_table = df_sims.to_html().replace("\\n", "<br>").replace("\\t", "&#09;")
    display(HTML(html_table))
    print("*****************************************")

# print (docs)

Similitudes para documento: 3449
label: talk.politics.guns
texto:

Funny, the medical examiner today stated that there was no
evidence ONE WAY or ANOTHER that there were bullet wounds --
not a single autopsy has been performed, so all reports are
deemed speculative.  INCLUDING reports that there were NO
bullet wounds.


Before long, I think all the kneejerk conspiracy theorists
are going to start getting pretty pissed off at how easily
they mislead themselves.  Also, pretty disappointed at
being ignored by the coutnry.



*****************************************
Similitudes para documento: 4228
label: misc.forsale
texto:
Hi netters,

Quantum LPS 240AT harddisk forsale.

3.5" frame, 1/3 height.
IDE format, master or slave
723 cyl 13 hd 51 s/t = 234.9 real megs
Access time of 16 ms.
256K cache on the drive

Asking $300. 


*****************************************
Similitudes para documento: 4627
label: rec.sport.baseball
texto:

Bill James is, however, very closely tied to STATS.


*****************************************
Similitudes para documento: 4506
label: talk.religion.misc
texto:
   Surely you are not equating David Koresh with Christianity? The two are
   not comparable.


*****************************************
Similitudes para documento: 4572
label: soc.religion.christian
texto:
[Someone quoted the following.  I've removed the name because it's not
clear which name goes with which level of quote.  --clh]



On the basis of these examples I would say that Joe Moore was only wrong
in claiming Augustine as a prime mover of the sin=sex view.  These quotes
clearly equate sexuality with defilement and incontinance, even within
the marriage relationship (else they would not apply to Mary after her marriage
to Joseph).

So Joe's assignment of the reasoning behind the concept of the perpetual
virginity of Mary does seem to be supported by these quotes.



*****************************************


#### Conclusiones

##### Documento a comparar 3449 con etiqueta talk.politics.guns  
El documento habla sobre una autopsia aparentemente por un homicidio que involucra armas y tiros.

Comparación con documento 1121, con similitud 0.461568 y etiqueta talk.politics.guns  
Coinciden las etiquetas. El texto tiene alguna similitud, hablan de autopsias, balas, heridas, evidencia, etc.

Comparación con documento 832, con similitud 0.317879 y etiqueta talk.religion.misc  
Las etiquetas son algo distintas, aunque tienen la misma base (talk).  
En cuanto al texto, también habla de reportes, balas. No parece un texto de religión, podría estar etiquetado como talk.politics.guns o algo mas parecido a esto.

Comparación con documento 4763, con similitud 0.245326 y etiqueta talk.politics.guns  
Coinciden las etiquetas. El texto habla sobre disparos, examen médico sobre el cuerpo (aparentemente un homicidio). Parece tener cierta similitud.

Comparación con documento 613, con similitud 0.239701 y etiqueta talk.politics.misc  
Etiquetas similares, pero no iguales, ambas son discusiones sobre politica, una relacionada a armas, otra a contenido general.
El texto también habla sobre cuerpos, disparos, heridas, tiene cierta similitud.

Comparación con documento 2695, con similitud 0.210677 y etiqueta talk.politics.misc  
Coinciden las etiquetas. El texto habla de examen médico sobre un cuerpo, evidencias, disparos, heridas, etc. También tiene cierta similitud.

##### Documento a comparar 4228 con etiqueta misc.forsale  
El documento habla sobre un aviso de venta de disco rígido.

Comparación con documento 4004, con similitud 0.227706 y etiqueta comp.sys.ibm.pc.hardware  
Etiquetas totalmente distintas. El texto habla sobre alguien que compró un disco, tiene cierta similitud ya que el documento comparado es sobre un anuncio de venta de un disco.

Comparación con documento 2991, con similitud 0.195262 y etiqueta comp.sys.ibm.pc.hardware  
Etiquetas totalmente distintas. En cuanto al texto, habla sobre problemas de configuración de un disco, lo cual tiene cierta similitud con el anuncio de venta. Ambos mencionan características del disco, como si funciona como master o slave, cache, formato IDE, etc.

Comparación con documento 6100, con similitud 0.184595 y etiqueta sci.electronics  
Etiquetas totalmente distintas. El texto hace recomendaciones sobre la compra de un disco IDE, con cache, etc, lo cual tiene cierta similitud.

Comparación con documento 4578, con similitud 0.183450 y etiqueta misc.forsale  
Coinciden las etiquetas. Tiene cierta similitud al ser otro anuncio de venta de disco, aunque al no describirlos de forma similar, pierde algo de similitud. A mi parecer, este sería un documento más similar que los anteriores, pero la métrica no dice lo mismo.

Comparación con documento 4100, con similitud 0.176129 y etiqueta comp.sys.ibm.pc.hardware  
Etiquetas totalmente distintas. El texto tiene referencia a un problema con discos en una pc, y la respuesta es una sugerencia sobre configuración de esos discos. No tiene mucha similitud con un aviso de venta en si, pero también hablan principalmente sobre discos rígidos.

##### Documento a comparar 4627 con etiqueta rec.sport.baseball  
El documento habla sobre un jugador de baseball y una empresa llamada STATS.

Comparación con documento 1968, con similitud 0.381607 y etiqueta rec.sport.baseball  
Coinciden las etiquetas. El texto es mucho mas largo, pero parece hablar sobre la misma discusión, el jugador y que es dueño de la empresa STATS, tiene mucha similitud en contenido general.

Comparación con documento 1231, con similitud 0.318910 y etiqueta rec.sport.baseball  
Coinciden las etiquetas. También se habla del mismo jugador, y de la empresa STATS, tiene similitud el contenido.

Comparación con documento 4968, con similitud 0.294406 y etiqueta rec.sport.baseball  
Coinciden las etiquetas. Otro documento donde se habla del jugador y la empresa, tiene similitud.

Comparación con documento 6174, con similitud 0.256514 y etiqueta rec.sport.baseball  
Coinciden las etiquetas. Este documento habla principalmente sobre la empresa STATS, no menciona al jugador. Tiene cierta similitud, pero no tanta como los documentos anteriores.

Comparación con documento 439, con similitud 0.208497 y etiqueta rec.sport.baseball  
Coinciden las etiquetas. Más allá de ser el 5to documento, en este se habla de estadísticas y no de la empresa, a simple vista no parece tener similitud en contenido interpretado.

##### Documento a comparar 4506 con etiqueta talk.religion.misc  
El documento habla sobre la comparación entre una persona y el cristianismo, diciendo que no son comparables.

Comparación con documento 6765, con similitud 0.418703 y etiqueta talk.religion.misc  
Coinciden las etiquetas. El texto tiene una cita directa al documento original, luego un comentario sobre la cita. el hecho de citar el documento original hace que tengan mucha similitud.

Comparación con documento 5363, con similitud 0.275627 y etiqueta talk.politics.misc  
Etiquetas distintas, aunque sobre la misma rama de discusión general. Se habla de la misma persona en otro contexto, la similitud esta dada por hablar de la misma persona.

Comparación con documento 2502, con similitud 0.259255 y etiqueta alt.atheism  
Etiquetas distintas, aunque puede haber alguna similitud al hablar de religión y ateismo. El documento menciona a la misma persona, no parece tener mucha similitud en contenido más allá de la mención.

Comparación con documento 1470, con similitud 0.243400 y etiqueta sci.crypt  
Etiquetas totalmente distintas. El texto es muy corto, no tiene similitud, la firma tiene el mismo nombre que la persona mencionada en el documento comparado, pero no tiene nada que ver una cosa con la otra.

Comparación con documento 4784, con similitud 0.234732 y etiqueta alt.atheism  
Etiquetas distintas, aunque como se mencionó en un caso anterior, puede haber similitudes. En cuanto al texto, el contenido no es similar, pero usa el apellido del personaje para hacer una comparación.

##### Documento a comparar 4572 con etiqueta soc.religion.christian  
El documento habla sobre pecado, sexualidad, impureza, matrimonio, e interpretación sobre como aplica a María y Jose.

Comparación con documento 3768, con similitud 0.299574 y etiqueta soc.religion.christian  
Coinciden las etiquetas. El texto tiene bastante relación con el pecado referido a María, e interpretaciones que lo hacen similar al documento comparado.

Comparación con documento 3672, con similitud 0.275061 y etiqueta soc.religion.christian  
Coinciden las etiquetas. El texto es mucho mas largo, pero habla también sobre María, el pecado, Dios, etc. Bastante relacionado con el texto comparado.

Comparación con documento 1716, con similitud 0.251366 y etiqueta soc.religion.christian  
Coinciden las etiquetas. Este texto es aun mas largo, pero en general habla principalmente sobre María, el pecado, la virginidad, temas que el documento comparado también tiene.

Comparación con documento 1051, con similitud 0.244472 y etiqueta soc.religion.christian  
Coinciden las etiquetas. Este documento no coincide casi nada, el tema principal es distinto, aunque habla de matrimonio, sexualidad, cristianismo, que usa como comparativas.

Comparación con documento 937, con similitud 0.237859 y etiqueta soc.religion.christian  
Coinciden las etiquetas. Este documento tiene alguna similitud, habla de Maria, el pecado, etc.